src/vsr/replica.zig

```diff
@@ -5254,26 +5290,59 @@ pub fn ReplicaType(
 
             self.set_op_and_commit_max(
                 do_view_change_head.op,
-                // `set_op_and_commit_max()` expects the highest commit_max that we know of.
-                // But DVCs include replica's `commit_min`, not `commit_max`.
-                std.math.max(
-                    self.commit_max,
-                    do_view_change_head.commit_min_max,
-                ),
+                commit_max: {
+                    var commit_max = self.commit_max;
+                    // `set_op_and_commit_max()` expects the highest commit_max that we know of.
+                    // But DVCs include replica's `commit_min`, not `commit_max`.
+                    commit_max = std.math.max(commit_max, do_view_change_head.commit_min_max);
+                    // An op cannot be uncommitted if it is definitely outside the pipeline.
+                    // Use `do_view_change_head.op` instead of `replica.op` since the former is
+                    // about to become the new `replica.op`.
+                    commit_max = std.math.max(
+                        commit_max,
+                        do_view_change_head.op -| constants.pipeline_max,
+                    );
+                    break :commit_max commit_max;
+                },
                 "on_do_view_change",
             );
             // "`replica.op` exists" invariant may be broken until after the canonical DVC headers
             // are installed.
 
-            // First, set all the canonical headers from the replica(s) with highest `view_normal`:
+            // "op_canonical" distinguishes between ops maybe-uncommitted ops from
+            // definitely-uncommitted ops — the latter are discarded to improve availablility.
+            // Gaps within a DVC's headers indicate ops that are not committed by that replica,
+            // but gaps in the new primary's headers cannot in general be truncated.
+            //
+            // Consider a view change with DVCs (pipeline_max=3):
+            //
+            //   replica   headers                       log_view
+            //         0   1 [2  3  4b]                  4     (new primary)
+            //         1   1  2  3  4a  5  6 [7  8  9]   5
+            //         2  (1  2  3  4a  5  6  7  8  9 )  5     (partitioned)
+            //
+            // Then replica 0 constructs:
+            //
+            //             1  2  3  4b        7  8  9
```

Comment:
This is such a killer bug. Nice find again @sentientwaffle !

If I understand correctly, the bug is as simple (and dangerous!) as a primary that is fairly far behind, simply receiving a DVC with headers further along in the log, then truncating everything in between.

Reply:
Yeah, though the `7 8 9` must come from the canonical view, and not have been committed by that replica.